mono/metadata/icall.c

```diff
@@ -6610,36 +6622,36 @@ mono_icall_get_environment_variable_names (MonoError *error)
 		++ n;
 
 	domain = mono_domain_get ();
-	names = mono_array_new_checked (domain, mono_defaults.string_class, n, error);
-	return_val_if_nok (error, NULL);
+	names = mono_array_new_handle (domain, mono_defaults.string_class, n, error);
+	if (!is_ok (error)) {
+		names = MONO_HANDLE_CAST (MonoArray, NULL_HANDLE);
+		goto leave;
+	}
 
 	n = 0;
 	for (e = environ; *e != 0; ++ e) {
 		parts = g_strsplit (*e, "=", 2);
 		if (*parts != 0) {
-			str = mono_string_new_checked (domain, *parts, error);
-			if (!is_ok (error)) {
+			if (!handle_environment_variable_array_set (domain, *parts, n, names, error)) {
 				g_strfreev (parts);
-				return NULL;
```

Comment:
We're no longer returning NULL if one string fails to construct?

Reply:
Partly agree with @Therzok, it's probably better to set `names` to `NULL` before `goto leave` here.  (It's not strictly necessary, because the caller: `ves_icall_System_Environment_GetEnvironmentVariableNames` will raise an exception in managed, rather than returning `null`).